File: Final_assembly/fire-alert-system/fire_alert_flutter/lib/screens/home_screen.dart
import 'package:flutter/material.dart';
import 'package:flutter_map/flutter_map.dart';
import 'package:latlong2/latlong.dart';
import '../services/api_service.dart';
import '../services/websocket_service.dart';
import '../utils/constants.dart';
import '../widgets/fire_point_marker.dart';
import 'escape_route_screen.dart';

class HomeScreen extends StatefulWidget {
  const HomeScreen({super.key});

  @override
  State<HomeScreen> createState() => _HomeScreenState();
}

class _HomeScreenState extends State<HomeScreen> {
  final MapController _mapController = MapController();
  List<Map<String, dynamic>> _firePoints = [];
  bool _loading = true;

  @override
  void initState() {
    super.initState();
    _loadFirePoints();
    _setupWebSocket();
  }

  void _setupWebSocket() {
    final wsService = WebSocketService();
    wsService.on('fire_point', _handleFirePoint);
    wsService.on('alert', _handleAlert);
    wsService.on('alert_resolved', _handleAlertResolved);
  }

  void _handleFirePoint(dynamic data) {
    setState(() {
      final index = _firePoints.indexWhere((p) => p['location'] == data['location']);
      if (index >= 0) {
        _firePoints[index] = {..._firePoints[index], ...data};
      } else {
        _firePoints.add(data);
      }
    });
  }

  void _handleAlert(dynamic data) {
    if (data['coordinates'] != null) {
      _handleFirePoint({
        'location': data['location'],
        'coordinates': data['coordinates'],
        'level': data['level'],
        'type': data['type'],
      });
    }
  }

  void _handleAlertResolved(dynamic data) {
    setState(() {
      _firePoints.removeWhere((p) => p['location'] == data['location']);
    });
  }

  Future<void> _loadFirePoints() async {
    try {
      final response = await ApiService.getActiveAlerts();
      if (response['success'] == true) {
        setState(() {
          _firePoints = (response['data'] as List)
              .map<Map<String, dynamic>>((alert) {
                    final coords = alert['coordinates'] ?? {};
                    return {
                      'id': alert['id'],
                      'location': alert['location'],
                      'latitude': coords['lat'] ?? coords['latitude'] ?? AppConstants.initialLatitude,
                      'longitude': coords['lng'] ?? coords['longitude'] ?? AppConstants.initialLongitude,
                      'level': alert['level'] ?? 'alert',
                      'type': alert['type'],
                    };
                  })
              .toList();
          _loading = false;
        });
      }
    } catch (error) {
      print('加载火点失败: $error');
      setState(() {
        _loading = false;
      });
    }
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Stack(
        children: [
          FlutterMap(
            mapController: _mapController,
            options: MapOptions(
              initialCenter: LatLng(AppConstants.initialLatitude, AppConstants.initialLongitude),
              initialZoom: 13.0,
            ),
            children: [
              TileLayer(
                urlTemplate: 'https://tile.openstreetmap.org/{z}/{x}/{y}.png',
                userAgentPackageName: 'com.firealertapp.fire_alert_app',
              ),
              MarkerLayer(
                markers: _firePoints
                    .map((point) => Marker(
                          point: LatLng(
                            point['latitude'] as double,
                            point['longitude'] as double,
                          ),
                          width: 40,
                          height: 40,
                          child: FirePointMarker(
                            level: point['level'] as String,
                          ),
                        ))
                    .toList(),
              ),
            ],
          ),
          Positioned(
            top: 10,
            left: 10,
            right: 10,
            child: Container(
              padding: const EdgeInsets.all(12),
              decoration: BoxDecoration(
                color: Colors.white.withOpacity(0.9),
                borderRadius: BorderRadius.circular(8),
                boxShadow: [
                  BoxShadow(
                    color: Colors.black.withOpacity(0.25),
                    blurRadius: 4,
                    offset: const Offset(0, 2),
                  ),
                ],
              ),
              child: Text(
                '活跃预警: ${_firePoints.length}',
                style: const TextStyle(
                  fontSize: 16,
                  fontWeight: FontWeight.w600,
                  color: Color(AppConstants.colorTextPrimary),
                ),
              ),
            ),
          ),
          if (_loading)
            const Center(
              child: CircularProgressIndicator(),
            ),
        ],
      ),
      floatingActionButton: FloatingActionButton.extended(
        onPressed: () {
          Navigator.push(
            context,
            MaterialPageRoute(
              builder: (context) => const EscapeRouteScreen(),
            ),
          );
        },
        icon: const Icon(Icons.directions_run),
        label: const Text('逃生路线'),
        backgroundColor: const Color(AppConstants.colorPrimary),
      ),
    );
  }
}

File: Final_assembly/fire-alert-system/mobile_app/src/screens/HomeScreen.js
/**
 * 首页 - 实时火点地图显示
 */

import React, { useState, useEffect } from 'react';
import {
  View,
  Text,
  StyleSheet,
  ActivityIndicator,
  Alert,
  TouchableOpacity
} from 'react-native';
import MapView, { Marker } from 'react-native-maps';
import { useTranslation } from 'react-i18next';
import { useNavigation } from '@react-navigation/native';
import Icon from 'react-native-vector-icons/MaterialIcons';
import WebSocketService from '../services/websocket';
import { getActiveAlerts } from '../services/api';
import { MAP_CONFIG, COLORS } from '../utils/constants';
import FirePointMarker from '../components/FirePointMarker';

const HomeScreen = () => {
  const { t } = useTranslation();
  const navigation = useNavigation();
  const [firePoints, setFirePoints] = useState([]);
  const [loading, setLoading] = useState(true);
  const [region, setRegion] = useState({
    latitude: MAP_CONFIG.INITIAL_LATITUDE,
    longitude: MAP_CONFIG.INITIAL_LONGITUDE,
    latitudeDelta: MAP_CONFIG.INITIAL_DELTA,
    longitudeDelta: MAP_CONFIG.INITIAL_DELTA,
  });

  useEffect(() => {
    // 加载初始数据
    loadFirePoints();

    // 订阅WebSocket事件
    WebSocketService.on('fire_point', handleFirePoint);
    WebSocketService.on('alert', handleAlert);
    WebSocketService.on('alert_resolved', handleAlertResolved);

    return () => {
      // 清理订阅
      WebSocketService.off('fire_point', handleFirePoint);
      WebSocketService.off('alert', handleAlert);
      WebSocketService.off('alert_resolved', handleAlertResolved);
    };
  }, []);

  const loadFirePoints = async () => {
    try {
      const response = await getActiveAlerts();
      if (response.success) {
        const points = response.data.map(alert => ({
          id: alert.id,
          location: alert.location,
          coordinates: alert.coordinates || { lat: region.latitude, lng: region.longitude },
          level: alert.level,
          type: alert.type,
          timestamp: alert.timestamp
        }));
        setFirePoints(points);
      }
    } catch (error) {
      console.error('加载火点失败:', error);
    } finally {
      setLoading(false);
    }
  };

  const handleFirePoint = (point) => {
    setFirePoints(prev => {
      const exists = prev.find(p => p.location === point.location);
      if (exists) {
        return prev.map(p =>
          p.location === point.location ? { ...p, ...point } : p
        );
      }
      return [...prev, point];
    });
  };

  const handleAlert = (alert) => {
    // 添加火点到地图
    if (alert.coordinates) {
      handleFirePoint({
        location: alert.location,
        coordinates: alert.coordinates,
        level: alert.level,
        type: alert.type,
        timestamp: alert.timestamp
      });
    }
  };

  const handleAlertResolved = (data) => {
    setFirePoints(prev => prev.filter(p => p.location !== data.location));
  };

  const handleMarkerPress = (point) => {
    Alert.alert(
      t('firePoints'),
      `${t('location')}: ${point.location}\n${t('status')}: ${t(point.level)}`,
      [{ text: t('resolveAlert'), onPress: () => {} }]
    );
  };

  if (loading) {
    return (
      <View style={styles.loadingContainer}>
        <ActivityIndicator size="large" color={COLORS.PRIMARY} />
        <Text style={styles.loadingText}>{t('realTimeMonitoring')}</Text>
      </View>
    );
  }

  return (
    <View style={styles.container}>
      <MapView
        style={styles.map}
        region={region}
        onRegionChangeComplete={setRegion}
      >
        {firePoints.map((point, index) => (
          <Marker
            key={point.id || index}
            coordinate={{
              latitude: point.coordinates.lat || point.coordinates.latitude,
              longitude: point.coordinates.lng || point.coordinates.longitude,
            }}
            onPress={() => handleMarkerPress(point)}
          >
            <FirePointMarker level={point.level} />
          </Marker>
        ))}
      </MapView>
      <View style={styles.infoContainer}>
        <Text style={styles.infoText}>
          {t('activeAlerts')}: {firePoints.length}
        </Text>
      </View>
      <TouchableOpacity
        style={styles.escapeRouteButton}
        onPress={() => navigation.navigate('EscapeRoute')}
      >
        <Icon name="directions-run" size={24} color="#fff" />
        <Text style={styles.escapeRouteButtonText}>{t('escapeRoute')}</Text>
      </TouchableOpacity>
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
  },
  map: {
    flex: 1,
  },
  loadingContainer: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
    backgroundColor: COLORS.BACKGROUND,
  },
  loadingText: {
    marginTop: 10,
    fontSize: 16,
    color: COLORS.TEXT_SECONDARY,
  },
  infoContainer: {
    position: 'absolute',
    top: 10,
    left: 10,
    right: 10,
    backgroundColor: 'rgba(255, 255, 255, 0.9)',
    padding: 12,
    borderRadius: 8,
    shadowColor: '#000',
    shadowOffset: { width: 0, height: 2 },
    shadowOpacity: 0.25,
    shadowRadius: 3.84,
    elevation: 5,
  },
  infoText: {
    fontSize: 16,
    fontWeight: '600',
    color: COLORS.TEXT_PRIMARY,
  },
  escapeRouteButton: {
    position: 'absolute',
    bottom: 20,
    left: 20,
    right: 20,
    backgroundColor: COLORS.PRIMARY,
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'center',
    padding: 16,
    borderRadius: 12,
    shadowColor: '#000',
    shadowOffset: { width: 0, height: 2 },
    shadowOpacity: 0.25,
    shadowRadius: 3.84,
    elevation: 5,
  },
  escapeRouteButtonText: {
    color: '#fff',
    fontSize: 18,
    fontWeight: '700',
    marginLeft: 8,
  },
});

export default HomeScreen;

